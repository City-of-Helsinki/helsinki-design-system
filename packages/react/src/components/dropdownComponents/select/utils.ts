import React from 'react';

import { SelectData, SelectMetaData, SelectDataHandlers, FilterFunction } from './types';
import {
  Group,
  Option,
  ModularOptionListData,
  ModularOptionListProps,
  ScreenReaderNotification,
} from '../modularOptionList/types';
import { ChangeEvent } from '../../dataProvider/DataContext';
import { eventTypes } from './events';
import {
  iterateAndCopyGroup,
  getGroupLabelOption,
  getAllOptions,
  propsToGroups,
  getSelectedOptions,
} from '../modularOptionList/utils';

// Re-export commonly used utility functions
export {
  getAllOptions,
  getSelectedOptions,
  propsToGroups,
  updateGroupLabelAndOptions,
  updateOptionInGroup,
  defaultFilter,
} from '../modularOptionList/utils';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export type OptionIterator = (
  option: Option,
  group: Group,
  optionIndex: number,
  groupIndex: number,
) => Option | undefined;

export const DROPDOWN_MENU_ITEM_HEIGHT = 52;

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = eventTypes.click, trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = eventTypes.change, trigger } = props;
  return {
    onChange: (originalEvent: React.ChangeEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}

export function getElementIds(containerId: string): SelectMetaData['elementIds'] {
  return {
    container: containerId,
    button: `${containerId}-main-button`,
    list: `${containerId}-list`,
    clearButton: `${containerId}-clear-button`,
    label: `${containerId}-label`,
    selectionsAndListsContainer: `${containerId}-sl-container`,
    tagList: `${containerId}-tag-list`,
    searchOrFilterInput: `${containerId}-input-element`,
    searchOrFilterInputLabel: `${containerId}-input-element-label`,
    clearAllButton: `${containerId}-clear-all-button`,
    showAllButton: `${containerId}-show-all-button`,
  };
}

export function countVisibleOptions(groups: SelectData['groups']): number {
  let count = 0;
  groups.forEach((group) => {
    group.options.forEach((option) => {
      if (!option.isGroupLabel && option.visible) {
        count += 1;
      }
    });
  });
  return count;
}

export function createScreenReaderNotification(type: string, content: string, delay = 0): ScreenReaderNotification {
  return {
    type,
    content,
    delay,
    showTime: 0,
    addTime: Date.now(),
  };
}

export function filterOptions(groups: ModularOptionListData['groups'], filterStr: string, filterFunc: FilterFunction) {
  const newGroupsWithFilteredOptions = iterateAndCopyGroup(groups, (option) => {
    if (option.isGroupLabel) {
      return { ...option };
    }
    return { ...option, visible: !filterStr || filterFunc(option, filterStr) };
  });

  // check if group label should be visible....
  newGroupsWithFilteredOptions.forEach((group) => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    const groupLabel = getGroupLabelOption(group);
    if (!groupLabel) {
      return;
    }
    groupLabel.visible = !!groupLabel.label && group.options.findIndex((opt) => !opt.isGroupLabel && opt.visible) > -1;
  });

  return newGroupsWithFilteredOptions;
}

export function filterSelectableOptions(
  groups: ModularOptionListData['groups'],
  filterStr: string,
  filterFunc: FilterFunction,
  isMultiSelect: boolean,
) {
  return getAllOptions(groups, !isMultiSelect).filter((option) => filterFunc(option, filterStr));
}

export function mergeSearchResultsToCurrent(
  props: Pick<ModularOptionListProps, 'groups' | 'options'>,
  currentGroups: ModularOptionListData['groups'],
): ModularOptionListData['groups'] {
  const newData = propsToGroups(props) || [];
  const newOptions = getAllOptions(newData);
  const currentOptionsWithoutMatches = getSelectedOptions(currentGroups).filter((option) => {
    const sameInNewOptionsIndex = newOptions.findIndex((newOption) => {
      return newOption.value === option.value;
    });
    if (sameInNewOptionsIndex > -1) {
      newOptions[sameInNewOptionsIndex].selected = true;
      return false;
    }
    return true;
  });

  const currentHiddenOptionsInAGroup = currentOptionsWithoutMatches.length
    ? [{ options: currentOptionsWithoutMatches.map((opt) => ({ ...opt, visible: false })) } as Group]
    : [];

  return [...currentHiddenOptionsInAGroup, ...newData];
}

/**
 *
 * @param notification
 * @param dataHandlers
 * @returns Returns true, if notification was added
 */
export function addOrUpdateScreenReaderNotificationByType(
  notification: ScreenReaderNotification,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfSameType = screenReaderNotifications.findIndex((n) => n.type === notification.type);
  if (indexOfSameType > -1) {
    const updatedList = [...screenReaderNotifications];
    const prev = updatedList[indexOfSameType];
    if (prev.content === notification.content) {
      return false;
    }
    updatedList[indexOfSameType] = notification;
    dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
    return false;
  }
  const updatedList = [...screenReaderNotifications, notification];
  dataHandlers.updateMetaData({ screenReaderNotifications: updatedList });
  return true;
}

export function removeScreenReaderNotification(
  target: Partial<ScreenReaderNotification>,
  dataHandlers: SelectDataHandlers,
) {
  const { screenReaderNotifications } = dataHandlers.getMetaData();
  const indexOfMatch = screenReaderNotifications.findIndex((n) => {
    const hasTypeMatch = !target.type || n.type === target.type;
    const hasContentMatch = !target.content || n.content === target.content;
    return hasTypeMatch && hasContentMatch;
  });
  if (indexOfMatch > -1) {
    screenReaderNotifications.splice(indexOfMatch, 1);
    dataHandlers.updateMetaData({ screenReaderNotifications });
    return true;
  }
  return false;
}
