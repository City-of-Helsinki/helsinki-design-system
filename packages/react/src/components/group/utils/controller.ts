import { Controller, ControllerProps } from '.';
import { createPropHandler } from './propHandler';
import { createStorage, StorageData } from './storage';

/**
 *
 * Controller for all Group's data and events.
 *
 */

export function createController({
  reRenderer,
  propSetter,
  initialData,
  onChange,
  metaData,
}: ControllerProps): Controller {
  const storage = createStorage(initialData);
  const meta = createStorage(metaData || {});
  const propHandler = createPropHandler(propSetter);
  return {
    getProps(props) {
      return propHandler.getProps({ ...props, controller: this });
    },
    getData() {
      return storage.get();
    },
    updateData({ data, updateKeys }) {
      storage.set(data);
      if (updateKeys && updateKeys.length) {
        updateKeys.forEach((id) => {
          propHandler.updateKey(id);
        });
      }
      reRenderer();
    },
    triggerChange(props) {
      return onChange({ ...props, controller: this });
    },
    getMetaData() {
      return meta.get() as StorageData;
    },
    updateMetaData(newData: StorageData) {
      return meta.set(newData);
    },
  };
}
