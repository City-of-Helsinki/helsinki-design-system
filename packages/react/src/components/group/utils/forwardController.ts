import React, { FC } from 'react';

import { Controller, HTMLElementAttributesWithChildren } from '.';

const HDSForwardController = Symbol('HDSForwardController');

/**
 *
 * Basically cloned how React.ForwardRef works.
 * Sometimes it is convenient to pass controller to a React.JSX/FC component.
 * If a React.JSX/FC component has HDSForwardController as a prop,
 * then the controller should be passed as a second argument.
 *
 */

export function ForwardController<P extends HTMLElementAttributesWithChildren & Record<string, unknown>>(
  component: (props: HTMLElementAttributesWithChildren & P, controller: Controller) => React.ReactElement | null,
): FC<HTMLElementAttributesWithChildren & P> {
  const renderFunc = (componentProps: P & { forwardedController?: Controller }) => {
    const { forwardedController, ...rest } = componentProps;
    if (!forwardedController) {
      throw new Error('forwardedController not passed to ForwardController');
    }
    return component((rest as unknown) as P, forwardedController);
  };
  Object.defineProperty(renderFunc, HDSForwardController, {
    value: true,
  });
  return renderFunc;
}

export function isForwardController(component: unknown) {
  const isFunctionOrObject = (checkTarget: unknown) => {
    // typeof null === object, so checking that too.
    const typeOfTarget = checkTarget ? typeof checkTarget : '';
    return typeOfTarget === 'object' || typeOfTarget === 'function';
  };
  const reactObjectType = isFunctionOrObject(component) && Reflect.get(component as React.ComponentType, 'type');
  return isFunctionOrObject(reactObjectType) && Reflect.get(reactObjectType, HDSForwardController) === true;
}
