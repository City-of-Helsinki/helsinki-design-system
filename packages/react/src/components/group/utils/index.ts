import {
  SyntheticEvent,
  FunctionComponent,
  HtmlHTMLAttributes,
  PropsWithChildren,
  ComponentType,
  MutableRefObject,
} from 'react';

import { StorageData, Storage } from './storage';

export type ChangeEventPayload = { value?: unknown; originalEvent?: SyntheticEvent };
export type ChangeEvent = { id: string; type?: string; payload?: ChangeEventPayload };
export type ChangeHandlerProps = ChangeEvent & { controller: Controller };
export type ChangeHandler = (props: ChangeHandlerProps) => unknown;
export type HTMLElementAttributesWithChildren = PropsWithChildren<HtmlHTMLAttributes<HTMLElement>>;
export type DefaultGroupElementProps = Record<string, unknown> & {
  key?: string;
  forwardedController?: Controller;
  'data-hds-group-id'?: string;
};

export type ElementRef = MutableRefObject<HTMLElement | null>;

export type Controller = {
  getProps: ({ id: string, elementProps: DefaultGroupElementProps }) => ReturnType<PropSetter>;
  updateData: (props: { data: StorageData; updateKeys?: string[] }) => void;
  getData: Storage['get'];
  getMetaData: () => StorageData;
  updateMetaData: (appended: StorageData) => StorageData;
  triggerChange: (change: ChangeEvent) => unknown;
};

export type ControllerProps = {
  reRenderer: () => void;
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange: ChangeHandler;
};

export type PropSetter<T = DefaultGroupElementProps> = (props: {
  id: string;
  controller: Controller;
  key: string;
  elementProps: DefaultGroupElementProps;
}) => T;

export type PropHandler = {
  getProps: (props: {
    id: string;
    controller: Controller;
    elementProps: DefaultGroupElementProps;
  }) => DefaultGroupElementProps;
  updateKey: (id: string) => void;
};

export type GroupChildFunction = FunctionComponent<{ controller: Controller }>;

export type GroupChild<P = DefaultGroupElementProps> = ComponentType<HTMLElementAttributesWithChildren & P>;

export type GroupProps = PropsWithChildren<unknown> & {
  propSetter: PropSetter;
  initialData: StorageData;
  metaData?: StorageData;
  onChange?: ChangeHandler;
};
