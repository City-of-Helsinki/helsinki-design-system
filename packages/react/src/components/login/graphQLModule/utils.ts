import { QueryOptions } from '@apollo/client';
import { merge } from 'lodash';

import { GraphQLModuleModuleProps } from '.';
import { ApiTokenClient } from '../apiTokensClient';

type PartialContext = Partial<QueryOptions['context']>;
type QueryHeaders = Record<string, unknown>;

function mergeQueryOptionsContexts(target: PartialContext, overrides: PartialContext): PartialContext {
  return merge(target, overrides);
}

export function mergeQueryOptionContexts(options: QueryOptions, overrideContext: PartialContext): QueryOptions {
  const context: PartialContext = options.context || {};
  // eslint-disable-next-line no-param-reassign
  options.context = mergeQueryOptionsContexts(context, overrideContext);
  return options;
}

export function mergeHeadersToQueryOptions(options: QueryOptions, headers: QueryHeaders): QueryOptions {
  const context: PartialContext = {
    headers,
  };
  return mergeQueryOptionContexts(options, context);
}

export function mergeAuthorizationHeaderToQueryOptions(options: QueryOptions, authHeader: string): QueryOptions {
  const headers: QueryHeaders = {
    authorization: authHeader,
  };
  return mergeHeadersToQueryOptions(options, headers);
}

export function mergeQueryOptions(
  target: Partial<GraphQLModuleModuleProps['queryOptions']>,
  overrides: Partial<GraphQLModuleModuleProps['queryOptions']>,
) {
  return merge(target, overrides);
}

/* No need to export to hds-react package */
export function mergeQueryOptionsToModuleProps<T, Q>(
  target: Partial<GraphQLModuleModuleProps<T, Q>>,
  queryOptions: Partial<QueryOptions<Q>>,
) {
  const { query, ...rest } = queryOptions;
  // eslint-disable-next-line no-param-reassign
  target.queryOptions = mergeQueryOptions(target.queryOptions || {}, rest);
  if (query) {
    // eslint-disable-next-line no-param-reassign
    target.query = query;
  }
  return target;
}

export function setBearerToQueryOptions(queryOptions: QueryOptions, token: string) {
  if (!token) {
    return queryOptions;
  }
  return mergeAuthorizationHeaderToQueryOptions(queryOptions, `Bearer ${token}`);
}

export function appendFetchOptions(queryOptions: QueryOptions, newOptions: Partial<Parameters<typeof fetch>[1]>) {
  if (!queryOptions.context) {
    // eslint-disable-next-line no-param-reassign
    queryOptions.context = {};
  }

  // eslint-disable-next-line no-param-reassign
  queryOptions.context.fetchOptions = {
    ...queryOptions.context.fetchOptions,
    ...newOptions,
  };
  return queryOptions;
}

/* No need to export to hds-react package */
export function mergeQueryOptionModifiers({
  options,
  queryHelper,
}: Pick<GraphQLModuleModuleProps, 'queryHelper' | 'options'>): Required<GraphQLModuleModuleProps>['queryHelper'] {
  const { apiTokenKey } = options || {};
  if (!apiTokenKey && !queryHelper) {
    return (opt: QueryOptions) => opt;
  }

  if (!apiTokenKey && queryHelper) {
    return queryHelper;
  }

  const tokenPicker = (apiTokenClient: ApiTokenClient, key: string) => {
    if (!apiTokenClient) {
      return undefined;
    }
    const tokens = apiTokenClient.getTokens();
    return tokens ? tokens[key] : undefined;
  };

  const apiTokenSetter: GraphQLModuleModuleProps['queryHelper'] = (opt, apiTokenClient) => {
    if (!apiTokenKey || !apiTokenClient) {
      return opt;
    }
    const token = tokenPicker(apiTokenClient, apiTokenKey);
    if (token) {
      return setBearerToQueryOptions(opt, token);
    }
    return opt;
  };

  if (!queryHelper) {
    return apiTokenSetter;
  }

  return (opt, apiTokenClient, beacon) => {
    return queryHelper(apiTokenSetter(opt, apiTokenClient, beacon), apiTokenClient, beacon);
  };
}
