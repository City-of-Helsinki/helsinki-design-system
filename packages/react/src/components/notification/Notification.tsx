import React, { useCallback, useEffect, useState } from 'react';
import { animated, useSpring } from '@react-spring/web';
import { VisuallyHidden } from '@react-aria/visually-hidden';

import '../../styles/base.module.css';
import styles from './Notification.module.css';
import classNames from '../../utils/classNames';
import { IconInfoCircleFill, IconErrorFill, IconAlertCircleFill, IconCheckCircleFill, IconCross } from '../../icons';
import { AllElementPropsWithoutRef } from '../../utils/elementTypings';

export type NotificationType = 'info' | 'error' | 'alert' | 'success';

export enum NotificationSize {
  Small = 'small',
  Medium = 'medium',
  Large = 'large',
}

export type NotificationPosition =
  | 'inline'
  | 'top-left'
  | 'top-center'
  | 'top-right'
  | 'bottom-left'
  | 'bottom-center'
  | 'bottom-right';

type CommonProps = React.PropsWithChildren<
  AllElementPropsWithoutRef<'section'> & {
    /**
     * Whether the notification should be closed automatically after a certain time
     * @default false
     */
    autoClose?: boolean;
    /**
     * The duration before the notification is automatically closed. Used together with the `autoClose` prop.
     * @default 6000
     */
    autoCloseDuration?: number;
    /**
     * Boolean indication whether notification has box shadow or not.
     */
    boxShadow?: boolean;
    /**
     * Additional class names to apply to the notification
     */
    className?: string;
    /**
     * Duration of the close fade-out animation
     * @default 85
     */
    closeAnimationDuration?: number;
    /**
     * Displays a progress bar on top of the notification when `true`
     * @default true
     */
    displayAutoCloseProgress?: boolean;
    /**
     * Determines whether the notification should be visually hidden. Useful when notification should only be "seen" by screen readers.
     * @default false
     */
    invisible?: boolean;
    /**
     * The label of the notification.
     * Note: Labels are not displayed visually for small notifications, but they are still accessible to assistive technology. This could be used to help screen reader users to better understand the context of the notification.
     */
    label?: string | React.ReactNode;
    /**
     * The aria-label of the notification region
     * @default "Notification"
     */
    notificationAriaLabel?: string;
    /**
     * Callback fired when the notification is closed
     */
    onClose?: () => void;
    /**
     * Override or extend the styles applied to the component
     */
    style?: React.CSSProperties;
    /**
     * The type of the notification
     * @default 'info'
     */
    type?: NotificationType;
    /**
     * The aria-level of the heading element
     * @default 2
     */
    headingLevel?: number;
  }
>;

type PositionAndSize =
  | { position?: 'inline'; size?: NotificationSize }
  | {
      /**
       * The position of the notification
       */
      position?: NotificationPosition;
      /**
       * The size of the notification
       */
      size?: Exclude<NotificationSize, NotificationSize.Large>;
    };

type Dismissible =
  | { dismissible?: false; closeButtonLabelText?: string }
  | {
      /**
       * Whether the notification can be closed
       * @default false
       */
      dismissible?: boolean;
      /**
       * The aria-label and title for the close button
       */
      closeButtonLabelText: string;
    };

export type NotificationProps = CommonProps & PositionAndSize & Dismissible;

// Icon mapping for notification types
const icons = {
  info: IconInfoCircleFill,
  success: IconCheckCircleFill,
  error: IconErrorFill,
  alert: IconAlertCircleFill,
};

/**
 * Returns the properties for the "open" transition
 * @param position
 */
const getOpenTransition = (position: NotificationPosition) => {
  const centerPositioned = position.includes('center');
  const animateFromTop = position.includes('top');

  return {
    from: {
      transform: `translate3d(${centerPositioned ? '-50%' : '0'}, ${animateFromTop ? '-' : ''}32px, 0)`,
      opacity: 0.66,
    },
    to: {
      transform: `translate3d(${centerPositioned ? '-50%' : '0'}, 0, 0)`,
      opacity: 1,
    },
    config: {
      friction: 30,
      tension: 300,
    },
  };
};

/**
 * Returns the properties for the "close" transition
 * @param duration
 */
const getCloseTransition = (duration: number) => ({
  from: { opacity: 1 },
  to: { opacity: 0 },
  config: { duration },
});

/**
 * Returns the properties for the "auto close" transition
 * @param duration
 */
const getAutoCloseTransition = (duration: number) => ({
  from: { transform: 'translate3d(0%, 0, 0)' },
  to: { transform: 'translate3d(-100%, 0, 0)' },
  config: {
    duration,
  },
});

/**
 * Conditionally hides the children visually, but leaves them accessible to assistive technology
 * @param visuallyHidden  Determines whether the children should be invisible
 * @param children
 * @constructor
 */
const ConditionalVisuallyHidden = ({ visuallyHidden, children }) =>
  visuallyHidden ? <VisuallyHidden>{children}</VisuallyHidden> : children;

export const Notification = React.forwardRef<HTMLDivElement, NotificationProps>(
  (
    {
      autoClose = false,
      autoCloseDuration = 6000,
      boxShadow = false,
      children,
      className = '',
      closeAnimationDuration = 85,
      closeButtonLabelText,
      dismissible = false,
      displayAutoCloseProgress = true,
      invisible = false,
      label,
      notificationAriaLabel = 'Notification',
      position = 'inline',
      onClose = () => null,
      size = NotificationSize.Medium,
      style,
      type = 'info',
      headingLevel = 2,
      ...rest
    }: NotificationProps,
    ref,
  ) => {
    const isToast: boolean = position !== 'inline';
    // only allow size 'large' for inline notifications
    if (isToast && size === NotificationSize.Large) {
      // eslint-disable-next-line no-console
      console.warn(`Size '${size}' is only allowed for inline positioned notifications`);
      // eslint-disable-next-line no-param-reassign
      size = NotificationSize.Medium;
    }
    // don't allow autoClose for inline notifications
    if (!isToast && autoClose) {
      // eslint-disable-next-line no-console
      console.warn(`The 'autoClose' property is not allowed for inline positioned notifications`);
      // eslint-disable-next-line no-param-reassign
      autoClose = false;
    }

    // internal state used for transitions
    const [open, setOpen] = useState(true);

    const handleClose = useCallback(() => {
      // trigger close animation
      setOpen(false);
      // emit onClose callback after the animation is completed
      setTimeout(() => onClose(), closeAnimationDuration);
    }, [onClose, closeAnimationDuration]);

    useEffect(() => {
      const interval = setTimeout(() => {
        if (autoClose) handleClose();
      }, autoCloseDuration);
      return () => clearTimeout(interval);
    }, [autoClose, autoCloseDuration, handleClose]);

    // icon
    const Icon = icons[type];

    // notification transitions
    const openTransitionProps = isToast ? getOpenTransition(position) : {};
    const closeTransitionProps = getCloseTransition(closeAnimationDuration);
    const autoCloseTransitionProps = displayAutoCloseProgress ? getAutoCloseTransition(autoCloseDuration) : {};

    const notificationTransition = useSpring(open ? openTransitionProps : closeTransitionProps);
    const autoCloseTransition = useSpring(autoCloseTransitionProps);

    return (
      <ConditionalVisuallyHidden visuallyHidden={invisible}>
        <animated.section
          // this "as" fixes ts error with wrong types. Seen only when running ts-check-stories.
          {...(rest as Record<string, unknown>)}
          // there is an issue with react-spring -rc3 and a new version of @types/react: https://github.com/react-spring/react-spring/issues/1102
          // eslint-disable-next-line  @typescript-eslint/no-explicit-any
          style={{ ...notificationTransition, ...(style as any) }}
          className={classNames(
            styles[position],
            styles.notification,
            styles[size],
            styles[type],
            autoClose && styles.noBorder,
            boxShadow && styles.boxShadow,
            className,
          )}
          aria-label={notificationAriaLabel}
          // Toast or invisible notifications require a role alert to ensure the screen readers will notify the content change.
          role={isToast || invisible ? 'alert' : undefined}
        >
          {autoClose && <animated.div style={autoCloseTransition} className={styles.autoClose} />}
          <div className={styles.content} ref={ref}>
            {label && (
              // Toast or invisible notifications do not always notice heading if role heading or aria-level is present.
              <div
                className={styles.label}
                {...(isToast || invisible ? {} : { role: 'heading', 'aria-level': headingLevel })}
              >
                <Icon className={styles.icon} />
                <ConditionalVisuallyHidden visuallyHidden={size === NotificationSize.Small}>
                  {label}
                </ConditionalVisuallyHidden>
              </div>
            )}
            {children && <div className={styles.body}>{children}</div>}
          </div>
          {dismissible && (
            <button
              className={classNames(styles.close, styles[type])}
              type="button"
              title={closeButtonLabelText}
              aria-label={closeButtonLabelText}
              onClick={handleClose}
            >
              <IconCross />
            </button>
          )}
        </animated.section>
      </ConditionalVisuallyHidden>
    );
  },
);
