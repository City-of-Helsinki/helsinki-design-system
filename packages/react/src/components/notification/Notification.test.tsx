import React, { HTMLAttributes } from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Notification, NotificationProps } from './Notification';
import { getCommonElementTestProps, getElementAttributesMisMatches } from '../../utils/testHelpers';

const label = 'This is the label text';
const body =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.';

describe('<Notification /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<Notification label={label}>{body}</Notification>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the component without optional label', () => {
    const { asFragment } = render(<Notification>{body}</Notification>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Native html props are passed to the element', async () => {
    const sectionProps: NotificationProps = getCommonElementTestProps('section');
    // the component has "notificationAriaLabel" prop
    sectionProps.notificationAriaLabel = sectionProps['aria-label'];
    const { getByTestId } = render(<Notification {...sectionProps}>{body}</Notification>);
    const element = getByTestId(sectionProps['data-testid']);
    expect(
      getElementAttributesMisMatches(element, {
        ...sectionProps,
        notificationAriaLabel: undefined,
      } as unknown as HTMLAttributes<HTMLElement>),
    ).toHaveLength(0);
  });

  // Replaces aria-atomic with role=alert and removes heading props for a better screen reader support
  it('uses toast notification props in non-inline notifications', () => {
    const { asFragment } = render(
      <Notification label={label} position="bottom-center">
        {body}
      </Notification>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(<Notification label={label}>{body}</Notification>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('adds className prop to notification classes', () => {
    const { container } = render(
      <Notification label={label} className="testClass">
        {body}
      </Notification>,
    );
    expect((container.firstChild as HTMLElement).classList.contains('testClass')).toBe(true);
  });

  it('should be closable', () => {
    const { getByRole } = render(
      <Notification label={label} dismissible closeButtonLabelText="test">
        {body}
      </Notification>,
    );
    expect(getByRole('button')).toBeDefined();
  });

  it('if headingLevel-property is not set, the aria-level of the heading element is 2 by default', () => {
    const { getByRole } = render(<Notification label={label}>{body}</Notification>);
    expect((getByRole('heading') as HTMLElement).getAttribute('aria-level')).toBe('2');
  });

  it('headingLevel-property sets the aria-level -attribute of the heading element', () => {
    const { getByRole } = render(
      <Notification label={label} headingLevel={3}>
        {body}
      </Notification>,
    );
    expect((getByRole('heading') as HTMLElement).getAttribute('aria-level')).toBe('3');
  });
});
