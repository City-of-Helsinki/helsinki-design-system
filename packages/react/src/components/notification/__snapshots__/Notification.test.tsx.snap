// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> spec adds role="alert" for non-inline notifications 1`] = `
<DocumentFragment>
  <section
    aria-atomic="true"
    aria-label="Notification"
    class="bottom-center notification default info"
    role="alert"
    style="transform: translate3d(-50%, 32px, 0); opacity: 0.66;"
  >
    <div
      class="content"
    >
      <div
        aria-level="2"
        class="label"
        role="heading"
      >
        <svg
          aria-hidden="true"
          class="icon s icon"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <rect
              height="24"
              width="24"
            />
            <path
              d="M12,3 C7.02943725,3 3,7.02943725 3,12 C3,16.9705627 7.02943725,21 12,21 C16.9705627,21 21,16.9705627 21,12 C21,7.02943725 16.9705627,3 12,3 Z M13,10 L13,16.5 L15,16.5 L15,18 L9,18 L9,16.5 L11,16.5 L11,11.5 L9,11.5 L9,10 L13,10 Z M11.8125239,6 C12.5373649,6 13.125,6.58759799 13.125,7.3125 C13.125,8.03735415 12.5373649,8.625 11.8125239,8.625 C11.0876351,8.625 10.5,8.03735415 10.5,7.3125 C10.5,6.58759799 11.0876351,6 11.8125239,6 Z"
              fill="currentColor"
            />
          </g>
        </svg>
        This is the label text
      </div>
      <div
        class="body"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`<Notification /> spec renders the component 1`] = `
<DocumentFragment>
  <section
    aria-atomic="true"
    aria-label="Notification"
    class="inline notification default info"
  >
    <div
      class="content"
    >
      <div
        aria-level="2"
        class="label"
        role="heading"
      >
        <svg
          aria-hidden="true"
          class="icon s icon"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <rect
              height="24"
              width="24"
            />
            <path
              d="M12,3 C7.02943725,3 3,7.02943725 3,12 C3,16.9705627 7.02943725,21 12,21 C16.9705627,21 21,16.9705627 21,12 C21,7.02943725 16.9705627,3 12,3 Z M13,10 L13,16.5 L15,16.5 L15,18 L9,18 L9,16.5 L11,16.5 L11,11.5 L9,11.5 L9,10 L13,10 Z M11.8125239,6 C12.5373649,6 13.125,6.58759799 13.125,7.3125 C13.125,8.03735415 12.5373649,8.625 11.8125239,8.625 C11.0876351,8.625 10.5,8.03735415 10.5,7.3125 C10.5,6.58759799 11.0876351,6 11.8125239,6 Z"
              fill="currentColor"
            />
          </g>
        </svg>
        This is the label text
      </div>
      <div
        class="body"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`<Notification /> spec renders the component without optional label 1`] = `
<DocumentFragment>
  <section
    aria-atomic="true"
    aria-label="Notification"
    class="inline notification default info"
  >
    <div
      class="content"
    >
      <div
        class="body"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.
      </div>
    </div>
  </section>
</DocumentFragment>
`;
